File: t/PACSFramework/VB/StorageServerManager/UI/ServerStatus/ServerInformationPresenter.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Leadtools.Dicom.Server.Admin
Imports System.ServiceProcess
Imports System.Threading
Imports System.Windows.Forms
Imports Leadtools.Demos.StorageServer.DataTypes
Imports Leadtools.Medical.Winforms
Imports System.Text.RegularExpressions
Imports System.Diagnostics
Imports System.Runtime.InteropServices
Imports System.ComponentModel

Namespace Leadtools.Demos.StorageServer.UI
   Friend Class ServerInformationPresenter
      Public Sub RunView(ByVal viewParam As ServerInformationView)
         View = viewParam

         __SyncContext = WindowsFormsSynchronizationContext.Current

         ConfigureView()

         AddHandler ServerState.Instance.ServerServiceChanged, AddressOf OnConfigureView
         AddHandler ServerState.Instance.IsRemoteServerChanged, AddressOf OnConfigureView
         AddHandler ServerState.Instance.RemoteServerInformationChanged, AddressOf OnConfigureView

         EventBroker.Instance.Subscribe(Of ServerSettingsAppliedEventArgs)(AddressOf OnConfigureView)
      End Sub

      Private Sub OnConfigureView(ByVal sender As Object, ByVal e As EventArgs)
         ConfigureView()
      End Sub

      Private Sub ConfigureViewDelegate(ByVal state As Object)
         If ServerState.Instance.IsRemoteServer Then
            View.CanStartStopServer = False

            If Not Nothing Is ServerState.Instance.RemoteServerInformation Then
               View.ServerDisplayName = ServerState.Instance.RemoteServerInformation.ServiceName
               View.ServerAE = ServerState.Instance.RemoteServerInformation.DicomServer.AE
               View.IpAddress = ServerState.Instance.RemoteServerInformation.DicomServer.IPAddress
                    View.Port = ServerState.Instance.RemoteServerInformation.DicomServer.Port.ToString()
                    View.IsSecure = ServerState.Instance.RemoteServerInformation.DicomServer.UseTls
                End If

            Return
         End If

         If Not Nothing Is Service Then
            If Service.Settings.Description.Length > 0 Then
               View.ServerDisplayName = Service.Settings.Description
            Else
               View.ServerDisplayName = "Storage Server"
            End If
            View.ServerAE = Service.Settings.AETitle
            View.IpAddress = Service.Settings.IpAddress.ToString()
                View.Port = Service.Settings.Port.ToString()
                View.IsSecure = Service.Settings.Secure

                AddHandler View.StartService, AddressOf view_StartService
            AddHandler View.StopService, AddressOf view_StopService

            SetServiceStatus()

            AddHandler Service.StatusChange, AddressOf Service_StatusChange
         Else
            View.ServerDisplayName = String.Empty
            View.ServerAE = String.Empty
            View.IpAddress = String.Empty
                View.Port = String.Empty
                View.IsSecure = False

                View.IsServerRunning = True
            View.CanStartStopServer = False
            View.Status = "Not Applicable"
         End If
      End Sub

      Private Sub ConfigureView()
         Dim service As DicomService = ServerState.Instance.ServerService

         '__SyncContext.Send(New SendOrPostCallback(AddressOf ConfigureViewDelegate), )
         Dim state As Object = New Object()
         __SyncContext.Send(AddressOf ConfigureViewDelegate, state)
      End Sub

      Sub view_StartService(ByVal sender As Object, ByVal e As EventArgs)
         Service.Start()
      End Sub

      Sub view_StopService(ByVal sender As Object, ByVal e As EventArgs)
         Service.Stop()
      End Sub

      Sub Service_StatusChange(ByVal sender As Object, ByVal e As EventArgs)
         SetServiceStatus()
      End Sub

      Private Sub MySendOrPostCallback(ByVal state As Object)
         If (Service.Status = ServiceControllerStatus.Running) Then
            View.IsServerRunning = True
         ElseIf Service.Status = ServiceControllerStatus.Stopped OrElse Service.Status = ServiceControllerStatus.StopPending OrElse Service.Status = ServiceControllerStatus.StartPending Then
            View.IsServerRunning = False
         Else
            View.CanStartStopServer = False
         End If

         '
         ' Convert CamelCase Status to a space delimeted status.  StartPending becomes Start Pending
         '
         View.Status = Regex.Replace(Service.Status.ToString(), "(\B[A-Z])", " $1")
      End Sub

      Private Sub SetServiceStatus()
         __SyncContext.Send(AddressOf MySendOrPostCallback, Nothing)
      End Sub

      Private _view As ServerInformationView
      Public Property View() As ServerInformationView
         Get
            Return _view
         End Get
         Private Set(ByVal value As ServerInformationView)
            _view = value
         End Set
      End Property

      Public ReadOnly Property Service() As DicomService
         Get
            Return ServerState.Instance.ServerService
         End Get
      End Property

      Private _mySyncContext As SynchronizationContext
      Private Property __SyncContext() As SynchronizationContext
         Get
            Return _mySyncContext
         End Get
         Set(ByVal value As SynchronizationContext)
            _mySyncContext = value
         End Set
      End Property
   End Class
end namespace 

File: t/PACSFramework/CS/StorageServerManager/UI/ServerStatus/ServerInformationPresenter.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Leadtools.Dicom.Server.Admin;
using System.ServiceProcess;
using System.Threading;
using System.Windows.Forms;
using Leadtools.Demos.StorageServer.DataTypes;
using Leadtools.Medical.Winforms;
using System.Text.RegularExpressions;
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.ComponentModel;
using System.Net;
using System.Net.NetworkInformation;
using Leadtools.Dicom.AddIn;

#if (LEADTOOLS_V19_OR_LATER_MEDICAL_VERIFY_ADDINS) || (LEADTOOLS_V19_OR_LATER)
using Leadtools.Medical.Winforms.ExternalStore;
using Leadtools.Dicom.AddIn.Common;
#endif

namespace Leadtools.Demos.StorageServer.UI
{
   class ServerInformationPresenter
   {
      private System.Timers.Timer _Timer = new System.Timers.Timer(500);
      private object _TimerLock = new object();

      public void RunView ( ServerInformationView view ) 
      {
         View = view ;
         
         __SyncContext = WindowsFormsSynchronizationContext.Current ;

         ConfigureView ( ) ;

         ServerState.Instance.ServerServiceChanged           += new EventHandler ( OnConfigureView ) ;
         ServerState.Instance.IsRemoteServerChanged          += new EventHandler ( OnConfigureView ) ;
         ServerState.Instance.RemoteServerInformationChanged += new EventHandler ( OnConfigureView ) ;
         
         EventBroker.Instance.Subscribe <ServerSettingsAppliedEventArgs> ( OnConfigureView ) ;         
         _Timer.Elapsed += new System.Timers.ElapsedEventHandler(_Timer_Elapsed);
         _Timer.Enabled = true;
      }

      void _Timer_Elapsed(object sender, System.Timers.ElapsedEventArgs e)
      {
         lock (_TimerLock)
         {
            try
            {
               _Timer.Stop();
               CheckIpAvailability();
            }
            catch { }
            _Timer.Start();
         }
      }      
      
      void OnConfigureView ( object sender, EventArgs e )
      {
         ConfigureView ( ) ;
      }

      private void ConfigureView ( )
      {
         DicomService service = ServerState.Instance.ServerService ;


         __SyncContext.Send ( new SendOrPostCallback ( delegate ( object state ) {
         
         if ( ServerState.Instance.IsRemoteServer )
         {
            View.CanStartStopServer = false ;
            
            if ( null != ServerState.Instance.RemoteServerInformation ) 
            {
               View.ServerDisplayName = ServerState.Instance.RemoteServerInformation.ServiceName ;
               View.ServerAE          = ServerState.Instance.RemoteServerInformation.DicomServer.AE ;
               View.IpAddress         = ServerState.Instance.RemoteServerInformation.DicomServer.IPAddress ;
               View.Port              = ServerState.Instance.RemoteServerInformation.DicomServer.Port.ToString ( ) ;
               View.IsSecure          = ServerState.Instance.RemoteServerInformation.DicomServer.UseTls;
            }
            
            return ;
         }
         
         if ( null != service ) 
         {
            View.ServerDisplayName = service.Settings.Description.Length>0?service.Settings.Description:"Storage Server" ;
            View.ServerAE          = service.Settings.AETitle ;
            View.IpAddress         = service.Settings.IpAddress.ToString ( ) ;
            View.Port              = service.Settings.Port.ToString ( ) ;
            View.IsSecure          = Service.Settings.Secure;

            View.StartService += new EventHandler ( view_StartService ) ;
            View.StopService  += new EventHandler ( view_StopService ) ;

            SetServiceStatus ( ) ;           
             
            Service.StatusChange += new EventHandler ( Service_StatusChange ) ;
            _Timer.Enabled = true;
         }
         else
         {
            _Timer.Enabled = false;
            View.ServerDisplayName = string.Empty ;
            View.ServerAE          = string.Empty ;
            View.IpAddress         = string.Empty ;
            View.Port              = string.Empty ;
            View.IsSecure          = false;

            View.IsServerRunning    = true ;
            View.CanStartStopServer = false ;
            View.Status             = "Not Applicable";               
         }
         } ), null ) ;
      }

      private void CheckIpAvailability()
      {
         IPGlobalProperties ipGlobalProperties = IPGlobalProperties.GetIPGlobalProperties();
         IPEndPoint[] endpoints = ipGlobalProperties.GetActiveTcpListeners();
         bool available = true;

         if (View.IsServerRunning)
            return;

         if (Service != null && Service.Settings != null)
         {
            foreach (IPEndPoint endpoint in endpoints)
            {
               if (endpoint.Address.ToString() == Service.Settings.IpAddress && endpoint.Port == Service.Settings.Port)
               {
                  available = false;
                  break;
               }
            }
         }

         if (!available && View.CanStart)
         {
            AsyncHelper.SynchronizedInvoke(View, () =>
            {
               View.CanStart = false;
               if (Service != null && Service.Settings != null)
               {
                  View.Port = Service.Settings.Port.ToString() + " [Port Unavailable]";
               }
            });
         }
         else if (available)
         {
            AsyncHelper.SynchronizedInvoke(View, () =>
           {
              View.CanStart = true;
              if (Service != null && Service.Settings != null)
              {
                 View.Port = Service.Settings.Port.ToString();
              }
           });
         }
      }

      void view_StartService ( object sender, EventArgs e )
      {
         _Timer.Enabled = false;
         if (Service != null)
         {
            Service.Start();
         }
      }

      void view_StopService(object sender, EventArgs e)
      {
         if (Service != null)
         {
            ServerState.Instance.ServerService.SendMessage(MessageNames.ServiceShuttingDown);
            Service.Stop();
         }
      }

      void Service_StatusChange ( object sender, EventArgs e )
      {
         if (Service.Status == ServiceControllerStatus.Stopped)
         {
            _Timer.Interval = 500;
            _Timer.Start();
         }

         SetServiceStatus ( ) ;
      }

#if (LEADTOOLS_V19_OR_LATER_MEDICAL_VERIFY_ADDINS) || (LEADTOOLS_V19_OR_LATER)
      private void QueryAddins()
      {
         if (ServerState.Instance.ServerService != null)
         {
            ServiceMessage message = new ServiceMessage();
            message.Message = MessageNames.IsAddinHealthy;
            message.Broadcast = true;

            try
            {
               // (05.23.2019)
               // If the administrative pipe is not connected, this can cause the StorageServerManager.exe UI to crash
               // I noticed this when the CSDicomHighLevelClient demo does a move, and I stop/start the CSStorageServerManger.exe listening service during the move
               // The UI will sometimes crash
               ServerState.Instance.ServerService.SendMessage(message);
            }
            catch (Exception)
            {
               // Console.WriteLine(ex.Message);
            }
         }
      }
#endif // #if (LEADTOOLS_V19_OR_LATER_MEDICAL_VERIFY_ADDINS) || (LEADTOOLS_V19_OR_LATER)

      private void SetServiceStatus ( )
      {
         string status = Regex.Replace(Service.Status.ToString(), "(\\B[A-Z])", " $1");

         __SyncContext.Send ( new SendOrPostCallback ( delegate ( object state ) {
            
                       
            View.IsServerRunning = Service.Status == ServiceControllerStatus.Running;
            View.CanStartStopServer = View.IsServerRunning ;                           

            //
            // Convert CamelCase Status to a space delimeted status.  StartPending becomes Start Pending
            //
            View.Status = status;

         } ), null ) ;

#if (LEADTOOLS_V19_OR_LATER_MEDICAL_VERIFY_ADDINS) || (LEADTOOLS_V19_OR_LATER)
         if (string.Equals(status, "Running",StringComparison.OrdinalIgnoreCase))
         {
            QueryAddins();
         }
#endif // #if (LEADTOOLS_V19_OR_LATER_MEDICAL_VERIFY_ADDINS) || (LEADTOOLS_V19_OR_LATER)
      }

      public ServerInformationView View
      {
         get ;
         private set ;
      }

      public DicomService Service
      {
         get 
         {
            return ServerState.Instance.ServerService ;
         }
      }

      private SynchronizationContext __SyncContext 
      {
         get ;
         set ;
      }
   }
}
